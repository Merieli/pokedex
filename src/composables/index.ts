export * from './useQuantityBasedOnWidth';
export * from './useSearchActions';
export * from './useValidateAndConvertHexColorToRgb';
