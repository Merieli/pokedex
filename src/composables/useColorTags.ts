import { computed } from 'vue';

import { colorTagSpecies } from '@/domain/character/colorTagSpecie';
import { UseColorTags } from '@/domain/character/UseColorTags.interface';

/**
 * Composable to get the color tag of the character
 * @param species - Species of the character
 * @param status - Status of the character
 * @returns {UseColorTags} - Computed values of the color tags
 */
export const useColorTags = (species: string, status?: string): UseColorTags => {
    const colorTagStatus = computed(() => {
        if (status === 'Dead') return 'red';

        if (status === 'Alive') return 'green';

        return 'gray';
    });

    const colorTagSpecie = computed(() => colorTagSpecies[species] || colorTagSpecies.Other);

    return {
        colorTagSpecie,
        colorTagStatus,
    };
};
