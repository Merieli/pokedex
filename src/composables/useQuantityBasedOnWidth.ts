import { computed, ComputedRef, ref } from 'vue';

export const useQuantityBasedOnWidth = () => {
    const widthWindow = ref(window.innerWidth);
    window.onresize = () => {
        widthWindow.value = window.innerWidth;
    };

    /**
     * Calculates the total load cards displayed for each screen width size
     */
    const totalOfCardsLoading: ComputedRef<number> = computed(() => {
        if (widthWindow.value > 976) return 16;

        if (widthWindow.value > 768) return 12;

        return 6;
    });

    /**
     * Calculates the total pages of the pagination component displayed according to the
     * screen width
     */
    const totalOfPagination: ComputedRef<number> = computed(() => {
        if (widthWindow.value > 1024) return 5;

        if (widthWindow.value > 768) return 4;

        return 3;
    });

    return {
        totalOfCardsLoading,
        totalOfPagination,
    };
};
