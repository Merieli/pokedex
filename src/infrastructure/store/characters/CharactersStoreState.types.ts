export interface CharactersStoreState {
    characters: any[];
}
