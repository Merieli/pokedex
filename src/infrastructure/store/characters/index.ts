export * from './CharactersStoreState.types';
export * from './StoreCharacters';
