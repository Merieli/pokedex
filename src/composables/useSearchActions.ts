import { computed, onBeforeMount } from 'vue';
import { useRoute, useRouter } from 'vue-router';

import { useCharactersStore } from '@/infrastructure/store/characters';

export const useSearchActions = () => {
    const route = useRoute();
    const router = useRouter();
    const store = useCharactersStore();

    const queryRoute = computed(() => route.query);

    /**
     * Searches for the desired name in the characters already loaded and saved in the store,
     * and saves it in the search characters list
     * @param currentName - current name of character
     */
    const searchByName = (currentName: string) => {
        router.push({});

        if (currentName && currentName.length >= 3) {
            store.$patch((state) => {
                state.isLoading = true;
                state.search.text = currentName;
            });

            const characters = store.allCharacters.filter((character) => {
                const textLowerCase = currentName.toLowerCase();
                const name: string = character.name.toLowerCase();
                return name.includes(textLowerCase);
            });

            setTimeout(() => {
                store.$patch((state) => {
                    state.search.characters = characters;
                    state.isLoading = false;
                });
            }, 3000);
        }
    };

    /**
     * Clears search term and search results
     */
    const clearSearch = () => {
        router.push({});
        store.$patch((state) => {
            state.isLoading = true;
            state.search.text = '';
            state.search.characters = [];
            state.random.character = [];
            state.random.show = false;
            state.isLoading = false;
        });
    };

    /**
     * Search for the character name in the api and save the list and the term searched
     * in the store
     */
    const searchInApi = async () => {
        router.push({ query: { character: store.search.text } });
        await store.findCharacterByName(store.search.text);
    };

    /**
     * Search for a random character in the api and save in the store
     */
    const searchRandomCharacter = async () => {
        router.push({ query: { random: 'true' } });
        await store.generateRandomCharacter();
    };

    /**
     * Updates the search term passed by parameter and performs the search in the api
     * @param current - current search term of route
     */
    const updateSearch = (current: string) => {
        if (store.allCharacters.length === 0) {
            store.$patch((state) => {
                state.search.text = current.toString();
            });

            searchInApi();
        }
    };

    /**
     * Get the query parameters and perform the search, when opened directly by route
     */
    const getQueryParamsAndLoadTheSearch = () => {
        if (queryRoute.value.character) {
            updateSearch(queryRoute.value.character.toString());
        }

        if (queryRoute.value.random === 'true') {
            searchRandomCharacter();
        }
    };

    onBeforeMount(getQueryParamsAndLoadTheSearch);

    return {
        searchByName,
        clearSearch,
        searchInApi,
        searchRandomCharacter,
    };
};
