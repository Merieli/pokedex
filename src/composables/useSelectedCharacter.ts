import { Character } from '@/domain';
import { useCharactersStore } from '@/infrastructure/store/characters';

export const useSelectedCharacter = () => {
    const store = useCharactersStore();

    const setsSelectedCharacter = (character: Character) => {
        store.$patch({
            selectedCharacter: {
                name: character.name,
                id: character.id,
                type: character.type,
                episode: character.episode,
                gender: character.gender,
                image: character.image,
                location: character.location,
                origin: character.origin,
                species: character.species,
                status: character.status,
            },
            isSelected: true,
        });
    };

    const clearSelectedCharacter = () => {
        store.$patch({
            isSelected: false,
            selectedCharacter: {
                id: '',
                name: '',
                episode: [],
                gender: 'unknown',
                status: 'unknown',
                type: '',
                image: '',
                species: 'unknown',
                location: {
                    id: '',
                    name: '',
                    dimension: '',
                    type: '',
                },
                origin: '',
            },
        });
    };

    return {
        setsSelectedCharacter,
        clearSelectedCharacter,
    };
};
