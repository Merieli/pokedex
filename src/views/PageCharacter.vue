<script setup lang="ts">
import { ComputedRef, computed, onBeforeMount, onUnmounted, toRefs } from 'vue';
import { useRoute, useRouter } from 'vue-router';

import CharacterTraitsTabs from '@components/CharacterTraitsTabs.vue';
import { Episode } from '@domain/index';

import { useColorTags } from '@/composables/useColorTags';
import { useSelectedCharacter } from '@/composables/useSelectedCharacter';
import { useCharactersStore } from '@/infrastructure/store/characters';

const store = useCharactersStore();
const router = useRouter();
const route = useRoute();
const { clearSelectedCharacter } = useSelectedCharacter();

const insertScrollIs = (isScroll: boolean) => {
    const { body } = document;

    body.classList[isScroll ? 'add' : 'remove']('fixed-scroll');
};

/**
 * Add a class to block scroll and if the character is not selected, it is selected by id from the
 * route
 */
onBeforeMount(async () => {
    insertScrollIs(true);

    if (!store.isSelected) {
        await store.selectCharacterById(Number(route.params.id));
    }
});

const { name, episode, gender, id, image, location, origin, species, status, type } = toRefs(store.selectedCharacter);

const lastEpisode: ComputedRef<Episode> = computed(() => episode.value[episode.value.length - 1]);

const { colorTagStatus, colorTagSpecie } = useColorTags(species.value, status.value || '');

const closeCharacter = () => {
    router.push({ name: 'home' });
    clearSelectedCharacter();
};

/**
 * Remove the class that blocks the scroll
 */
onUnmounted(() => insertScrollIs(false));
</script>

<template>
    <v-card v-if="store.isSelected && !store.isLoading" class="character-traits">
        <v-card-title class="character-traits__title">
            <div class="character-traits__button-wrapper">
                <v-btn
                    class="character-traits__back-button"
                    variant="text"
                    icon="mdi-arrow-left"
                    data-character-traits="button"
                    @click="closeCharacter"
                ></v-btn>
            </div>
            <h3 class="character-traits__name" data-character-traits="name">{{ name }}</h3>
            <h4 class="character-traits__id" data-character-traits="id"># {{ id }}</h4>
        </v-card-title>
        <v-card-subtitle class="character-traits__tags">
            <v-chip
                class="character-traits__tag"
                :color="colorTagSpecie"
                variant="elevated"
                data-character-traits="specie"
            >
                {{ species }}
            </v-chip>
            <v-chip
                v-if="status"
                class="character-traits__tag"
                :color="colorTagStatus"
                variant="elevated"
                data-character-traits="status"
            >
                {{ status }}
            </v-chip>
        </v-card-subtitle>
        <v-card-text>
            <img class="character-traits__image" :src="image" alt="" />
        </v-card-text>
        <CharacterTraitsTabs
            :gender="gender"
            :origin="origin"
            :type="type"
            :location="location"
            :last-episode="lastEpisode"
            :episode="episode"
        ></CharacterTraitsTabs>
    </v-card>
</template>

<style lang="postcss" scoped>
.character-traits {
    @apply min-w-full h-full
        bg-meri-light;

    &__title {
        @apply flex justify-start items-center flex-wrap;
    }

    &__title,
    &__tags {
        @apply pl-10 pr-10
            max-w-5xl mx-auto;
    }

    &__tag {
        @apply mr-2 mb-2;
    }

    &__name {
        @apply font-extrabold;
    }

    &__id {
        @apply font-semibold text-base
            ml-4;
    }

    &__button-wrapper {
        @apply w-full
            flex justify-start;
    }

    &__image {
        @apply rounded-3xl  max-w-[230px] tablet:max-w-xs m-auto;
    }
}
</style>
<style>
.fixed-scroll {
    position: fixed;
    width: 100%;
    overflow-y: hidden;
    scrollbar-width: none;
    -ms-overflow-style: none;
}
.fixed-scroll::-webkit-scrollbar {
    display: none;
    width: 0;
}

.character-traits__overlay .v-overlay__content {
    width: 100%;
    height: 100vh;
}
</style>
