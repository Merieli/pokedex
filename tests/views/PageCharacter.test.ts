import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { Pinia } from 'pinia';
import { createVuetify } from 'vuetify';
import { mockSelectedCharacter } from '../__mocks__/mockSelectedCharacter';

import { useSelectedCharacter } from '@/composables/useSelectedCharacter';
import { useCharactersStore } from '@/infrastructure/store/characters';
import PageCharacter from '@/views/PageCharacter.vue';

describe('PageCharacter.vue', () => {
    let pinia: Pinia;

    const mockRoute = {
        params: {
            id: 1,
        },
    };

    const mockRouter = {
        push: vi.fn(),
    };

    vi.mock('@/composables/useSelectedCharacter');

    const setupWrapper = () => {
        pinia = createTestingPinia({
            initialState: {
                characters: {
                    selectedCharacter: mockSelectedCharacter,
                    isSelected: true,
                },
            },
        });

        const vuetify = createVuetify();

        return {
            wrapper: mount(PageCharacter, {
                global: {
                    plugins: [pinia, vuetify],
                    stubs: {
                        teleport: true,
                    },
                    mocks: {
                        $route: mockRoute,
                        $router: mockRouter,
                    },
                },
            }),
            store: useCharactersStore(),
        };
    };

    describe('Integração/Componente', () => {
        describe('👀 Renderização:', () => {
            test('Dado o componente Quando renderizado Então deve possuir html equivalente ao Snapshot gravado', () => {
                const { wrapper } = setupWrapper();

                expect(wrapper.html()).toMatchSnapshot();
            });

            test('Dado uma lista de características do personagem Quando renderizado Então deve exibir o nome e id do personagem', () => {
                const { wrapper } = setupWrapper();
                const characterName = wrapper.find('[data-character-traits="name"]');
                const characterId = wrapper.find('[data-character-traits="id"]');

                expect(characterName.text()).toBe(mockSelectedCharacter.name);
                expect(characterId.text()).toBe(`# ${mockSelectedCharacter.id}`);
            });

            test('Dado uma lista de caraterísticas do personagem Quando renderizado Então deve exibir as tags de espécie e estado de vida', () => {
                const { wrapper } = setupWrapper();
                const characterSpecie = wrapper.find('[data-character-traits="specie"]');
                const characterStatus = wrapper.find('[data-character-traits="status"]');

                expect(characterSpecie.text()).toBe(mockSelectedCharacter.species);
                expect(characterStatus.text()).toBe(mockSelectedCharacter.status);
            });
        });
        describe('🧠 Comportamento:', () => {
            test('Dado a lista de características Quando clicar no botão de voltar e a store remover o usuário selecionado Então deve ', async () => {
                const { wrapper, store } = setupWrapper();
                const spyUpdateStore = vi.spyOn(useSelectedCharacter, 'clearSelectedCharacter');
                const buttonBack = wrapper.find('[data-character-traits="button"]');

                await buttonBack.trigger('click');

                expect(useSelectedCharacter).toHaveBeenCalledTimes(1);
                expect(spyUpdateStore.mock.calls[0][0]).toContain({ isSelected: false });
            });
        });

        describe('🐕 Navegação:', () => {
            test('Dado a pagina de personagem Quando clicar no botao voltar Então deve ir para pagina inicial', async () => {
                const { wrapper, store } = setupWrapper();
                const buttonBack = wrapper.find('[data-character-traits="button"]');

                await buttonBack.trigger('click');
            });
        });
    });
});
