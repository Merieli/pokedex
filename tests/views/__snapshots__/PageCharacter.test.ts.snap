// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageCharacter.vue > Integra√ß√£o/Componente > üëÄ Renderiza√ß√£o: > Dado o componente Quando renderizado Ent√£o deve possuir html equivalente ao Snapshot gravado 1`] = `
"<v-card class=\\"character-traits\\" data-v-bc83ff91=\\"\\">
  <v-card-title class=\\"character-traits__title\\" data-v-bc83ff91=\\"\\">
    <div class=\\"character-traits__button-wrapper\\" data-v-bc83ff91=\\"\\">
      <v-btn class=\\"character-traits__back-button\\" variant=\\"text\\" icon=\\"mdi-arrow-left\\" data-character-traits=\\"button\\" data-v-bc83ff91=\\"\\"></v-btn>
    </div>
    <h3 class=\\"character-traits__name\\" data-character-traits=\\"name\\" data-v-bc83ff91=\\"\\">Rick Sanchez</h3>
    <h4 class=\\"character-traits__id\\" data-character-traits=\\"id\\" data-v-bc83ff91=\\"\\"># 1</h4>
  </v-card-title>
  <v-card-subtitle class=\\"character-traits__tags\\" data-v-bc83ff91=\\"\\">
    <v-chip class=\\"character-traits__tag\\" color=\\"#f2a38b\\" variant=\\"elevated\\" data-character-traits=\\"specie\\" data-v-bc83ff91=\\"\\">Human</v-chip>
    <v-chip class=\\"character-traits__tag\\" color=\\"green\\" variant=\\"elevated\\" data-character-traits=\\"status\\" data-v-bc83ff91=\\"\\">Alive</v-chip>
  </v-card-subtitle>
  <v-card-text data-v-bc83ff91=\\"\\"><img class=\\"character-traits__image\\" src=\\"https://rickandmortyapi.com/api/character/avatar/1.jpeg\\" alt=\\"\\" data-v-bc83ff91=\\"\\"></v-card-text>
  <character-traits-tabs-stub gender=\\"Male\\" origin=\\"Earth (C-137)\\" type=\\"Human normal\\" location=\\"[object Object]\\" lastepisode=\\"[object Object]\\" episode=\\"[object Object],[object Object],[object Object]\\" data-v-bc83ff91=\\"\\"></character-traits-tabs-stub>
</v-card>"
`;
